src/flags.zig

```diff
@@ -0,0 +1,127 @@
+
+/// Format and print an error message followed by the usage string to stderr,
+/// then exit with an exit code of 1.
+pub fn fatal(comptime fmt_string: []const u8, args: anytype) noreturn {
+    const stderr = std.io.getStdErr().writer();
+    stderr.print("error: " ++ fmt_string ++ "\n", args) catch {};
+    std.os.exit(1);
+}
+
+pub fn parse_commands(args: *std.process.ArgIterator, comptime Commands: type) Commands {
+    assert(@typeInfo(Commands) == .Union);
+
+    var subcommand = args.next() orelse fatal("subcommand required", .{});
+    inline for (comptime std.meta.fields(Commands)) |field| {
+        comptime assert(std.mem.indexOf(u8, field.name, "_") == null);
+        if (std.mem.eql(u8, subcommand, field.name)) {
+            return @unionInit(Commands, field.name, parse_flags(args, field.field_type));
+        }
+    }
+    fatal("unknown subcommand: '{s}'", .{subcommand});
+}
+
+pub fn parse_flags(args: *std.process.ArgIterator, comptime Flags: type) Flags {
+    if (Flags == void) {
+        if (args.next()) |arg| {
+            fatal("unexpected argument: '{s}'", .{arg});
+        }
+        return {};
+    }
+
+    assert(@typeInfo(Flags) == .Struct);
+
+    var result: Flags = undefined;
+    var counts: FlagsCounts(Flags) = .{};
+
+    next_arg: while (args.next()) |arg| {
+        inline for (comptime std.meta.fields(Flags)) |field| {
+            const flag = comptime field_to_flag(field.name);
+            if (std.mem.startsWith(u8, arg, flag)) {
+                const flag_value = parse_flag(flag, arg);
+
+                @field(counts, field.name) += 1;
+                if (@field(counts, field.name) > 1) {
+                    fatal("{s}: duplicate argument", .{flag});
+                }
+
+                @field(result, field.name) = parse_int(field.field_type, flag, flag_value);
+                continue :next_arg;
+            }
+        }
+
+        fatal("unexpected argument: '{s}'", .{arg});
+    }
+
+    inline for (comptime std.meta.fields(Flags)) |field| {
+        if (@field(counts, field.name) == 0) {
+            if (field.default_value) |default| {
+                @field(result, field.name) = @ptrCast(
+                    *const field.field_type,
+                    @alignCast(std.meta.alignment(field.field_type), default),
+                ).*;
+            } else {
+                const flag = comptime field_to_flag(field.name);
+                fatal("{s}: argument is required", .{flag});
+            }
+        }
+    }
+
+    return result;
+}
+
+fn field_to_flag(comptime field: []const u8) []const u8 {
+    comptime var result: []const u8 = "--";
+    comptime {
+        var index = 0;
+        while (std.mem.indexOf(u8, field[index..], "_")) |i| {
+            result = result ++ field[index..][0..i] ++ "-";
+            index = index + i + 1;
+        }
+        result = result ++ field[index..];
+    }
+    return result;
+}
+
+fn FlagsCounts(comptime Flags: type) type {
```

Comment:
```suggestion
fn FlagsCountsType(comptime Flags: type) type {
```

We've started doing this pattern in the forest. It would be good to add to TigerStyle too:

Suffix type constructor names with `Type`. This enables the resulting type to be given the name that the constructor might otherwise have shadowed.